digitaltwin_client/src/dt_interface_client.c

```diff
@@ -278,65 +287,184 @@ static bool DT_IsAlpha(char c)
             (c >= 'A' && c <= 'Z'));
 }
 
-// DT_InterfaceClient_CheckNameValid makes sure that the passed names (specifically on registration) is legal.
-// In general the SDK does NOT do policy checking like this, leaving it to the server.  We make an exception
-// in this case because an ill-formatted values on registration over MQTT will result in the
-// TCP connection being closed.  Because this can be so hard for developer to diagnose interface being
-// invalid versus network issues, we provide this check.
-int DT_InterfaceClient_CheckNameValid(const char* valueToCheck, bool isInterfaceId)
+// IsEndOfSegmentCharacter returns whether the current character is the end of a segment.
+// For component names, this will simply be a 0.  For interfaces, it will be a ';' or ':'.
+static bool IsEndOfSegmentCharacter(char c, DT_SEGMENT_TYPE segmentType)
+{
+    if (segmentType == DT_SEGMENT_TYPE_COMPONENT_NAME)
+    {
+        return (c==0);
+    }
+    else
+    {
+        return (c == DT_Segment_Separator || c == DT_Version_Separator);
+    }
+}
+
+// IsNameSegmentValid checks whether either a component name or a segment of
+// an interface is legal DTMI.  Legal requires the segment to start with an alphabetical character,
+// be followed by alpha/numbers/_, but not end with a _.
+static bool IsNameSegmentValid(const char* segmentName, DT_SEGMENT_TYPE segmentType, const char** endPosition)
```

Comment:
This can't be done in a regex expression?

Reply:
It can, but the formal definition of this is quite long.  This is one I thought the English (and of course code) was easier to follow.

For languages with regex parsers this would presumably how they'd chech string and logic below would be in code and self-documenting.

```
dtmi =       "dtmi" ":" dtpath ";" dtver
dtpath =     dtseg *(":" dtseg)
dtseg =      dtuserseg / dssysseg
dtuserseg =  ALPHA [ *(ALPHA / DIGIT / "_") (ALPHA / DIGIT) ]
dtsysseg =   "_" *(ALPHA / DIGIT / "_") (ALPHA / DIGIT)
dtver =      digitnz 0*8DIGIT
digitnz =    %x31-39
             ; 1-9
```